app/src/screen.c

```diff
@@ -454,6 +466,90 @@ update_texture(struct screen *screen, const AVFrame *frame) {
     }
 }
 
+bool
+screen_render_text(struct screen *screen) {
+    struct render_text_request req;
+
+    // we check this flag outside lock, it may cause that we don't show text immediately.
+    // but the performance not affected.
+    if (!screen->need_render_text) {
+        return true;
+    }
+
+    if (screen->curr_text.text == NULL) {
+        mutex_lock(screen->mutex);
+        if (cbuf_is_empty(&screen->queue)) {
+            screen->need_render_text = false;
+            mutex_unlock(screen->mutex);
+            return true;
+        }
+        cbuf_take(&screen->queue, &req);
+        mutex_unlock(screen->mutex);
+        screen->curr_text.text = req.text;
+        screen->curr_text.time = req.time;
+    }
+
+    bool ret = false;
+    SDL_RWops *raw_font = open_sans_font_raw();
+    TTF_Font *font = TTF_OpenFontRW(raw_font, 1, 25);
+    SDL_Color text_color = {0, 0, 0, 255};
+    SDL_Color bg_color = {255, 255, 255, 50};
+    SDL_Surface *surface = NULL;
+    SDL_Texture *text_texture = NULL;
+    int width = 0;
+    int height = 0;
+    SDL_Rect dstrect;
+
+    if (!font) {
+        LOGW("Init font failed: %s", SDL_GetError());
+        goto end;
+    }
+
+    surface = TTF_RenderText_Shaded(font, screen->curr_text.text, text_color, bg_color);
+    if (!surface) {
+        LOGW("Init surface failed: %s", SDL_GetError());
+        goto end;
+    }
+
+    text_texture = SDL_CreateTextureFromSurface(screen->renderer, surface);
+    if (!text_texture) {
+        LOGW("Init text texture failed: %s", SDL_GetError());
+        goto end;
+    }
+
+    if (TTF_SizeText(font, screen->curr_text.text, &width, &height)) {
+        LOGW("Could not get the size of TTF: %s", TTF_GetError());
+        goto end;
+    }
+
+    dstrect.x = screen->rect.w / 2 - width / 2;
+    dstrect.y = screen->rect.h / 3 * 2;
+    dstrect.w = width;
+    dstrect.h = height;
+
+    SDL_RenderCopy(screen->renderer, text_texture, NULL, &dstrect);
+
+    screen->curr_text.time -= 1;
+    if (screen->curr_text.time <= 0) {
+        SDL_free(screen->curr_text.text);
+        screen->curr_text.text = NULL;
+        screen->curr_text.time = 0;
```

Comment:
The time is handled as the number of frames. But the framerate is (very) variable: it may be as low as 0 fps if the device screen is constant to 90fps (for example) if it changes a lot.

The "toast" delay should be based on real time instead.

Reply:
Agree.

Reply:
What should happen if time is up but the screen is constant? In correct implementation, it will remain on screen until screen changed.

Reply:
IMO, the popup/taost delay should be independant of the device video. It should fade out after a constant delay (whether the device screen is constant or not).